pkg/agent/cache/manager.go

```diff
@@ -0,0 +1,372 @@
+	BaseRegEntries []*proto.RegistrationEntry
+	Logger         logrus.FieldLogger
+}
+
+type manager struct {
+	managedCache     Cache
+	serverCerts      []*x509.Certificate
+	serverAddr       string
+	serverSPIFFEID   string
+	reason           error
+	entryRequestCh   chan map[string][]EntryRequest
+	regEntriesCh     chan []*proto.RegistrationEntry
+	cacheEntryCh     chan CacheEntry
+	spiffeIdEntryMap map[string]CacheEntry
+	baseSVID         []byte
+	baseSVIDKey      *ecdsa.PrivateKey
+	baseSPIFFEID     string
+	baseRegEntries   []*proto.RegistrationEntry
+	log              logrus.FieldLogger
+	ctx              context.Context
+	cancel           context.CancelFunc
+	doneCh           chan struct{}
+}
+
+func NewManager(ctx context.Context, c *MgrConfig) (Manager, error) {
+
+	cert, err := x509.ParseCertificate(c.BaseSVID)
+	if err != nil {
+		return nil, err
+	}
+	basespiffeID, err := uri.GetURINamesFromCertificate(cert)
+	if err != nil {
+		return nil, err
+	}
+	ctx, cancel := context.WithCancel(ctx)
+
+	return &manager{
+		managedCache:     NewCache(),
+		serverCerts:      c.ServerCerts,
+		serverSPIFFEID:   c.ServerSPIFFEID,
+		serverAddr:       c.ServerAddr,
+		baseSVID:         c.BaseSVID,
+		baseSVIDKey:      c.BaseSVIDKey,
+		baseSPIFFEID:     basespiffeID[0],
+		baseRegEntries:   c.BaseRegEntries,
+		log:              c.Logger.WithField("subsystem_name", "cacheManager"),
+		spiffeIdEntryMap: make(map[string]CacheEntry),
+		entryRequestCh:   make(chan map[string][]EntryRequest),
+		regEntriesCh:     make(chan []*proto.RegistrationEntry),
+		cacheEntryCh:     make(chan CacheEntry),
+		ctx:              ctx,
+		cancel:           cancel,
+		doneCh:           make(chan struct{}),
+	}, nil
+}
+
+func (m *manager) Shutdown(err error) {
+	m.log.Debug("Shutting Down Cache Manager ", err)
+	m.reason = err
+	m.cancel()
+}
+func (m *manager) Done() <-chan struct{} {
+	return m.doneCh
+}
+func (m *manager) Err() (err error) {
+	<-m.Done()
+	return m.reason
+}
+func (m *manager) Cache() Cache {
+	return m.managedCache
+}
+
+func (m *manager) Init() {
+	defer close(m.doneCh)
+	var wg sync.WaitGroup
+	wg.Add(1)
+	go m.expiredCacheEntryHandler(5*time.Second, &wg)
+
+	wg.Add(1)
+	go m.regEntriesHandler(&wg)
+	m.log.Debug("Initializing Cache Manager", " baseSVIDSPIFFEId:", m.baseSPIFFEID)
+
+	m.regEntriesCh <- m.baseRegEntries
+
+	for {
+
+		var svid []byte
+		var key *ecdsa.PrivateKey
+		select {
+		case reqs := <-m.entryRequestCh:
+			for parentId, entryRequests := range reqs {
+				wg.Add(1)
+				if _, ok := m.spiffeIdEntryMap[parentId]; ok {
+					svid = m.spiffeIdEntryMap[parentId].SVID.SvidCert
+					key = m.spiffeIdEntryMap[parentId].PrivateKey
+				}
+				if parentId == m.baseSPIFFEID {
+					svid = m.baseSVID
+					key = m.baseSVIDKey
+				}
+				conn, err := m.getGRPCConn(svid, key)
+				if err != nil {
+					m.Shutdown(err)
+					break
+				}
+				m.log.Debug("Spawning FetchID for", "entryRequests:", entryRequests)
```

Comment:
Do we need spaces at the end of these strings? Also, is `entryRequests` cleanly printable? I saw some hex debug output during your demo yesterday. Another option is to use structured logging. Similar comment about many of the other debug statements in here